src/core/server/saved_objects/migrationsv2/actions/index.ts

```diff
@@ -419,6 +420,132 @@ export const pickupUpdatedMappings = (
+  return client
+    .openPointInTime({
+      index,
+      keep_alive: pitKeepAlive,
+    })
+    .then((response) => Either.right({ pitId: response.body.id }))
+    .catch(catchRetryableEsClientErrors);
+};
+
+/** @internal */
+export interface ReadWithPit {
+  outdatedDocuments: SavedObjectsRawDoc[];
+  readonly lastHitSortValue: number[] | undefined;
+}
+
+/*
+ * Requests documents from the index using PIT mechanism.
+ * Filter unusedTypesToExclude documents out to exclude them from being migrated.
+ * */
+export const readWithPit = (
+  client: ElasticsearchClient,
+  pitId: string,
+  /* When reading we use a source query to exclude saved objects types which
+   * are no longer used. These saved objects will still be kept in the outdated
+   * index for backup purposes, but won't be available in the upgraded index.
+   */
+  unusedTypesQuery: Option.Option<estypes.QueryContainer>,
+  batchSize: number,
+  searchAfter?: number[]
+): TaskEither.TaskEither<RetryableEsClientError, ReadWithPit> => () => {
+  return client
+    .search<SavedObjectsRawDoc>({
+      body: {
+        // Sort fields are required to use searchAfter, so we set some defaults here
+        sort: {
+          updated_at: { order: 'desc' },
+        },
+        pit: { id: pitId, keep_alive: pitKeepAlive },
+        size: batchSize,
+        search_after: searchAfter,
+        // Exclude saved object types
+        query: Option.fold<estypes.QueryContainer, estypes.QueryContainer | undefined>(
+          () => undefined,
+          (query) => query
+        )(unusedTypesQuery),
+      },
+    })
+    .then((response) => {
+      const hits = response.body.hits.hits;
+
+      if (hits.length > 0) {
+        return Either.right({
+          // @ts-expect-error @elastic/elasticsearch _source is optional
+          outdatedDocuments: hits as SavedObjectsRawDoc[],
+          lastHitSortValue: hits[hits.length - 1].sort as number[],
+        });
+      }
+
+      return Either.right({
+        outdatedDocuments: [],
+        lastHitSortValue: undefined,
+      });
+    })
+    .catch(catchRetryableEsClientErrors);
+};
+
+/*
+ * Closes PIT.
+ * See https://www.elastic.co/guide/en/elasticsearch/reference/current/point-in-time-api.html
+ * */
+export const closePit = (
+  client: ElasticsearchClient,
+  pitId: string
+): TaskEither.TaskEither<RetryableEsClientError, {}> => () => {
+  return client
+    .closePointInTime({
+      body: { id: pitId },
+    })
+    .then((response) => {
+      if (!response.body.succeeded) {
+        throw new Error(`Failed to close PointInTime with id: ${pitId}`);
+      }
+      return Either.right({});
+    })
+    .catch(catchRetryableEsClientErrors);
+};
+
+/*
+ * Transform outdated docs and write them to the index.
+ * */
+export const transformDocs = (
+  client: ElasticsearchClient,
+  transformRawDocs: TransformRawDocs,
+  outdatedDocuments: SavedObjectsRawDoc[],
+  index: string,
+  refresh: estypes.Refresh
```

Comment:
I'd move to named arguments instead, but let's do it in a separate PR for all the actions at once